### PR TITLE
Add fast paths to matrix sum function

### PR DESCRIPTION
**Pull request by jonreiter**
Add fast paths to matrix sum function for triangular, symmetric and vector.

There was a TODO in there and I call this a lot on symmetric matrices.

As this is partially a performance-related change I added benchmarks for sum.
The new version is faster for triangular and symmetric -- but as the
benchmark is new I'm not sure a clean way to benchstat this without two
pull requests.

Also add a few function comments to triangular.go to appease golint.

Please take a look.

### CODE REVIEW

File: mat/dense_test.go [hunk 1]
```diff
@@ -2093,3 +2093,14 @@ func denseMulTransSymBench(b *testing.B, size int, rho float64) {
 		wd = &n
 	}
 }
+
+func BenchmarkDenseSum1000(b *testing.B) { denseSumBench(b, 1000, 100) }
```
Comment (kortschak) [inline mat/dense_test.go]: This looks like a job for sub-benchmarks.

File: mat/dense_test.go [hunk 2]
```diff
@@ -2093,3 +2093,14 @@ func denseMulTransSymBench(b *testing.B, size int, rho float64) {
...
+
+func denseSumBench(b *testing.B, size, iters int) {
+	b.StopTimer()
+	a, _ := randDense(size, 1.0, rand.NormFloat64)
+	b.StartTimer()
+	for i := 0; i < iters; i++ {
```
Comment (kortschak) [inline mat/dense_test.go]: This is not correct. The form of a benchmark loop is `for i := 0; i < b.N; i++ {`

File: mat/dense_test.go [hunk 3]
```diff
@@ -2093,3 +2093,14 @@ func denseMulTransSymBench(b *testing.B, size int, rho float64) {
...
+		_ = Sum(a)
```
Comment (kortschak) [inline mat/dense_test.go]: This should be assigned to a value that is used or a global.

File: mat/matrix.go [hunk 4]
```diff
@@ -848,20 +848,57 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
...
 	aU, _ := untranspose(a)
-	if rma, ok := aU.(RawMatrixer); ok {
+	switch rma := aU.(type) {
+	case RawSymmetricer:
+		rm := rma.RawSymmetric()
+		for i := 0; i < rm.N; i++ {
+			// Diagonals count once while off-diagonals count twice.
+			sum += rm.Data[i*rm.Stride+i]
+			for _, v := range rm.Data[i*rm.Stride+i+1 : i*rm.Stride+rm.N] {
+				sum += 2 * v
```
Comment (kortschak) [inline mat/matrix.go]: Do a single doubling at the end.

File: mat/matrix.go [hunk 5]
```diff
@@ -848,20 +848,57 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
...
+		return sum
+	case RawMatrixer:
 		rm := rma.RawMatrix()
 		for i := 0; i < rm.Rows; i++ {
 			for _, v := range rm.Data[i*rm.Stride : i*rm.Stride+rm.Cols] {
 				sum += v
 			}
 		}
 		return sum
+	case *VecDense:
```
Comment (kortschak) [inline mat/matrix.go]: This isn't mentioned in the commit message. Please do.

File: mat/matrix.go [hunk 6]
```diff
@@ -848,20 +848,57 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
...
+		rm := rma.RawVector()
+		for i := 0; i < rm.N; i++ {
+			sum += rm.Data[i*rm.Inc]
+		}
+		return sum
 	}
+	// Finally the generic case.
```
Comment (kortschak) [inline mat/matrix.go]: Maybe rather than this comment, just make a `default:` case in the switch.

File: mat/symmetric_test.go [hunk 7]
```diff
@@ -735,3 +735,28 @@ func TestPowPSD(t *testing.T) {
 		}
 	}
 }
+
+func BenchmarkSymSum1000(b *testing.B) { symSumBench(b, 1000, 100) }
```
Comment (kortschak) [inline mat/symmetric_test.go]: See previous comments about sub-benchmarks and benchmark iterations.

File: mat/triangular.go [hunk 8]
```diff
@@ -105,6 +105,7 @@ func (t TransposeTri) Untranspose() Matrix {
 	return t.Triangular
 }
 
+// UntransposeTri and return specifically a Triangular object.
```
Comment (kortschak) [inline mat/triangular.go]: These documentation changes are incidental and should be in a separate PR.

File: mat/triangular_test.go [hunk 9]
```diff
@@ -512,3 +512,25 @@ func TestCopySymIntoTriangle(t *testing.T) {
 		}
 	}
 }
+
+func BenchmarkTriSum1000(b *testing.B) { triSumBench(b, 1000, 100) }
```
Comment (kortschak) [inline mat/triangular_test.go]: See previous benchmark comments.

File: mat/dense_test.go (hunk 3 above)
Comment (sbinet) [inline mat/dense_test.go]: one should probably store the result in a global variable, to prevent the clever optimizer from doing its job.

File: mat/dense_test.go (hunk 2 above)
Comment (sbinet) [inline mat/dense_test.go]: shouldn't this be: `s/iters/b.N/` instead?

File: mat/dense_test.go (hunk 3 above)
Comment (sbinet) [reply mat/dense_test.go]: ie:
```go
var denseSumBenchResult float64

func denseSumBench(b *testing.B, size int) {
    a, _ := randDense(size, 1.0, rand.NormFloat64)
    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        denseSumBenchResult = Sum(a)
    }
}
```

File: mat/symmetric_test.go [hunk 10]
```diff
@@ -735,3 +735,28 @@ func TestPowPSD(t *testing.T) {
...
+
+func symSumBench(b *testing.B, size, iters int) {
+	b.StopTimer()
+	a := randSymDense(size)
+	b.StartTimer()
+	for i := 0; i < iters; i++ {
```
Comment (sbinet) [inline mat/symmetric_test.go]: ditto. (`b.N` + global variable)

File: mat/triangular_test.go [hunk 11]
```diff
@@ -512,3 +512,25 @@ func TestCopySymIntoTriangle(t *testing.T) {
...
+
+func triSumBench(b *testing.B, size, iters int) {
+	b.StopTimer()
+	a := randTriDense(size)
+	b.StartTimer()
+	for i := 0; i < iters; i++ {
```
Comment (sbinet) [inline mat/triangular_test.go]: ditto.

**Comment by jonreiter** [comment]
benchstat:

```
name                old time/op  new time/op   delta
VectorSum100000-4   515µs ± 1%   106µs ± 1%   -79.44%  (p=0.029 n=4+4)
TriSum1000-4        5.55ms ± 1%  0.61ms ± 1%  -88.96%  (p=0.029 n=4+4)
SymSum1000-4        6.71ms ± 9%  0.61ms ± 1%  -90.85%  (p=0.029 n=4+4)
DenseSum1000-4      1.09ms ± 0%  1.09ms ± 1%     ~     (p=0.343 n=4+4)

```

File: mat/triangular_test.go [hunk 12]
```diff
@@ -512,3 +512,27 @@ func TestCopySymIntoTriangle(t *testing.T) {
...
+	a := randTriDense(size)
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		triSumForBench = Sum(a)
+	}
+}
+
+func randTriDense(size int) *TriDense {
+	t := NewTriDense(size, Upper, nil)
+	return t
```
Comment (sbinet) [inline mat/triangular_test.go]: debug statement leftover?

File: mat/vector_test.go [hunk 13]
```diff
@@ -561,3 +561,16 @@ func randVecDense(size, inc int, rho float64, rnd func() float64) *VecDense {
...
 }
+
+func BenchmarkVectorSum100000(b *testing.B) { vectorSumBench(b, 100000) }
+
+var vectorSumForBench float64
+
+func vectorSumBench(b *testing.B, size int) {
+	b.StopTimer()
+	a := randVecDense(size, 1, 1.0, rand.NormFloat64)
+	b.StartTimer()
```
Comment (sbinet) [inline mat/vector_test.go]: I still find using `b.RestartTimer()` easier (and dropping `b.StopTimer()` entirely) but ok.

File: mat/dense_test.go (hunk 1 above)
Comment (jonreiter) [reply mat/dense_test.go]: i think i've addressed all the benchmark issues.

File: mat/matrix.go (hunk 4 above)
Comment (jonreiter) [reply mat/matrix.go]: now breaks out a single doubling and does the diagonal.  neither approach looks totally clean to me honestly.

File: mat/triangular.go (hunk 8 above)
Comment (jonreiter) [reply mat/triangular.go]: ive just removed them.

File: mat/triangular_test.go (hunk 12 above)
Comment (jonreiter) [reply mat/triangular_test.go]: indeed. sorted now. thanks spot.

File: mat/matrix.go [hunk 14]
```diff
@@ -848,26 +848,65 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
...
 			}
 		}
 		return sum
-	}
-	for i := 0; i < r; i++ {
-		for j := 0; j < c; j++ {
-			sum += a.At(i, j)
+	case *VecDense:
+		rm := rma.RawVector()
+		for i := 0; i < rm.N; i++ {
```
Comment (kortschak) [inline mat/matrix.go]: This is not correct. `rm.Inc` is not guaranteed to be 1.

File: mat/matrix.go [hunk 15]
```diff
@@ -848,26 +848,65 @@ func normLapack(norm float64, aTrans bool) lapack.MatrixNorm {
...
 	var sum float64
 	aU, _ := untranspose(a)
-	if rma, ok := aU.(RawMatrixer); ok {
+	switch rma := aU.(type) {
+	case RawSymmetricer:
+		rm := rma.RawSymmetric()
+		for i := 0; i < rm.N; i++ {
+			// Diagonals count once while off-diagonals count twice.
+			sum += rm.Data[i*rm.Stride+i]
+			subSum := 0.0
```
Comment (kortschak) [inline mat/matrix.go]: `var subSum float64` (though I'd just call this `s`).

File: mat/matrix.go (hunk 14 above)
Comment (jonreiter) [reply mat/matrix.go]: isn't rm.Inc the spacing between elements per http://www.netlib.org/lapack/explore-html/de/d05/dasum_8f_source.html (for example)?

Comment (jonreiter) [reply mat/matrix.go]: i think i have rebased etc now per your request (i have never done that before so hopefully correct).

**Comment by kortschak** [comment]
What I was suggest was the following:

Start an interactive rebase with 8 commits.
```
$ git rebase -i HEAD~8 # 8 is the number of commits in the PR plus one because we are modifying the first one.
```


This will start an editor (whichever editor is you default - usually vim) that will show you this (with some additional commentary/instructions below it thanks to `git`)
```
pick 3d26580 lapack/testlapack: replace all literal constants in dgesvdTest with tol expressions
pick 9f02923 Add fast paths to matrix sum function for triangular, symmetric and vector. There was a TODO in there and I call this a lot on symmetric matrices.
pick 3f5a65c Revert "Add fast paths to matrix sum function for triangular, symmetric and vector."
pick 3208802 mat: add benchmarks for Sum with Dense, SymDense, TriDense and VecDense
pick 80d1b4c mat: add fast paths for Sum with SymDense, TriDense and VecDense
pick 32192d5 A+C: add Jonathan Reiter
pick 20ad6d0 remove mistaken debug statement
pick f35ed94 small cleanups
```

You then change it to look like this (the commented stuff below it is not impotrant)
```
pick 3d26580 lapack/testlapack: replace all literal constants in dgesvdTest with tol expressions


pick 3208802 mat: add benchmarks for Sum with Dense, SymDense, TriDense and VecDense
squash 20ad6d0 remove mistaken debug statement
pick 80d1b4c mat: add fast paths for Sum with SymDense, TriDense and VecDense
pick 32192d5 A+C: add Jonathan Reiter
pick f35ed94 small cleanups
```
and then write out the file and close it.

It will come back with a file buffer that has this text.
```
# This is a combination of 2 commits.
# The first commit's message is:

mat: add benchmarks for Sum with Dense, SymDense, TriDense and VecDense

# This is the 2nd commit message:

remove mistaken debug statement
```

Edit it to look like thie
```
mat: add benchmarks for Sum with Dense, SymDense, TriDense and VecDense
```

and then writ it out and close it.

This will leave you with a new history which looks like this:
```
d0ac098 small cleanups
c928e9b A+C: add Jonathan Reiter
9045fee mat: add fast paths for Sum with SymDense, TriDense and VecDense
c920e60 mat: add benchmarks for Sum with Dense, SymDense, TriDense and VecDense
```

(with different shas).

Ideally, I would like the "cleanups" commit to be squashed as wel, but it has changes corresponding to two different commits (the benchmarks and the substantive changes).
This can be fixed though; if you `git reset --mixed HEAD~1` after the procedure above, you'll have that commit back as a set of changes that you can merge as two separate commits, and then do the same kind of interactive rebase as above, squashing each to the appropriate place.

If anything goes wrong during this, you can always just `git reset --hard f35ed94` to take it back to the state it's at now.

When it's all good (you can check by doing `git diff master origin/master` — assuming your remotes is origin — and confirm no difference), just `git push -f origin master`.

File: mat/matrix.go (hunk 4 above)
Comment (kortschak) [reply mat/matrix.go]: No, but this has fewer multiplies.

**Comment by jonreiter** [comment]
I think correct now.  Reread a few times...again sorry if still not quite there.

**Comment by jonreiter** [comment]
@kortschak 

I think correct now. Reread a few times...again sorry if still not quite there.

File: mat/matrix.go (hunk 14 above)
Comment (kortschak) [reply mat/matrix.go]: Sorry, you are right.

**Comment by kortschak** [comment]
This is almost there, but I don't see an easy rebase to get to where we want to be (it's possible, but it's much easier to do a mixed reset to the commit before you started and do the commits again). We want changes to the benchmarks in the first commit, changes to the `Sum` code in the second (no changes to the benchmarks here) and then your addition to A+C in the third. The final state is OK.

Thank you for bearing with me on this.
